ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::ffi::c_void;
+use std::iter::once;
+
+use deno_core::v8::fast_api;
+use dynasmrt::dynasm;
+use dynasmrt::DynasmApi;
+use dynasmrt::ExecutableBuffer;
+
+use crate::NativeType;
+use crate::Symbol;
+
+pub(crate) fn is_compatible(sym: &Symbol) -> bool {
+  cfg!(all(target_arch = "x86_64", target_family = "unix"))
+    && !sym.can_callback
+    && !sym.parameter_types.iter().any(|t| !is_fast_api_arg(*t))
+    && is_fast_api_rv(sym.result_type)
+}
+
+pub(crate) fn compile_trampoline(sym: &Symbol) -> Trampoline {
+  Trampoline::compile(sym)
+}
+
+pub(crate) fn make_template(sym: &Symbol, trampoline: &Trampoline) -> Template {
+  let args = once(fast_api::Type::V8Value)
+    .chain(sym.parameter_types.iter().map(|t| t.into()))
+    .collect::<Vec<_>>();
+
+  Template {
+    args: args.into_boxed_slice(),
+    ret: (&fast_api::Type::from(&sym.result_type)).into(),
+    symbol_ptr: trampoline.ptr(),
+  }
+}
+
+/// Trampoline for fast-call FFI functions
+///
+/// Removes first argument (Javascript object receiver) and shifts the rest of arguments to the left
+pub(crate) struct Trampoline(ExecutableBuffer);
+
+impl Trampoline {
+  fn ptr(&self) -> *const c_void {
+    &self.0[0] as *const u8 as *const c_void
+  }
+
+  fn compile(sym: &Symbol) -> Self {
+    #[derive(Clone, Copy)]
+    struct ArgCounter {
+      integer: i32,
+      sse: i32,
+    }
+    let mut counter = ArgCounter {
+      integer: -6,
+      sse: -8,
+    };
+    let args_to_shift: Vec<_> = sym
+      .parameter_types
+      .iter()
+      .filter_map(|p| match p {
+        NativeType::F32 | NativeType::F64 => {
+          counter.sse += 1;
+          if counter.sse > 0 && counter.integer >= 0 {
+            // floats are only shifted to accomodate integer shift in the stack
+            Some(counter)
+          } else {
+            None
+          }
+        }
+        _ => {
+          counter.integer += 1;
+          Some(counter)
+        }
+      })
+      .collect();
+
+    let mut ops = dynasmrt::x64::Assembler::new().unwrap();
+    if args_to_shift.is_empty() {
+      dynasm!(ops
+        ; .arch x64
+        ; xor rdi, rdi
```

Comment:
At least I would love to see relatively heavy comments on these instructions.

eg. Why is it that rdi is xor'd on empty arguments?

Reply:
> At least I would love to see relatively heavy comments on these instructions.

Absolutely. I'll document the important aspects of the calling convention. 

>  Why is it that rdi is xor'd on empty arguments?

To erase the receiver from the first argument. One could declare the function without arguments but actually expect it in the function implementation. I can imagine that it is important to protect against such access, right? Actually I intent to write a test for such case. Any suggestion on what harm could be exemplified in the test with such concealed access?

Reply:
If it matters, we do rather blindly assume that the user provided the proper routine type signature.

Reply:
Hmm... Currently I think the receiver is always just unused on both ends. Node somehow hides in a return value ArrayBuffer I think? But I think it's a good idea to clean it out, just to be consistent.

Reply:
Hmmm. Actually seems like compilers don't generally clear out the extra parameter: https://godbolt.org/z/q8WTacscE

So maybe we don't need to, either?

Reply:
yeah, but normally compilers optimize for performance, not security. That's why crates like [zeroize](https://docs.rs/zeroize/latest/zeroize/) have to exist.

I'm thinking along the lines of some form of prototype pollution for example. I prefer to err on the side of caution in this case, but not my call.

Reply:
Can mark this resolved.